datafusion/core/src/physical_optimizer/join_selection.rs

```
@@ -0,0 +1,972 @@
+        )
+    });
+    let right_len = right_cols.len();
+    let left_cols = left_schema.fields().iter().enumerate().map(|(i, f)| {
+        (
+            Arc::new(Column::new(f.name(), right_len + i)) as Arc<dyn PhysicalExpr>,
+            f.name().to_owned(),
+        )
+    });
+
+    left_cols.chain(right_cols).collect()
+}
+
+/// Swaps join sides for filter column indices and produces new JoinFilter
+fn swap_join_filter(filter: &Option<JoinFilter>) -> Option<JoinFilter> {
+    match filter {
+        Some(filter) => {
+            let column_indices = filter
+                .column_indices()
+                .iter()
+                .map(|idx| {
+                    let side = if matches!(idx.side, JoinSide::Left) {
+                        JoinSide::Right
+                    } else {
+                        JoinSide::Left
+                    };
+                    ColumnIndex {
+                        index: idx.index,
+                        side,
+                    }
+                })
+                .collect();
+
+            Some(JoinFilter::new(
+                filter.expression().clone(),
+                column_indices,
+                filter.schema().clone(),
+            ))
+        }
+        None => None,
+    }
+}
+
+impl PhysicalOptimizerRule for JoinSelection {
+    fn optimize(
+        &self,
+        plan: Arc<dyn ExecutionPlan>,
+        session_config: &SessionConfig,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        let collect_left_threshold = session_config.hash_join_collect_left_threshold;
+        plan.transform_up(&|plan| {
+            if let Some(hash_join) = plan.as_any().downcast_ref::<HashJoinExec>() {
+                match hash_join.partition_mode() {
+                    PartitionMode::Auto => {
+                        try_collect_left(hash_join, Some(collect_left_threshold))
+                            .unwrap()
+                            .or_else(|| Some(partitioned_hash_join(hash_join).unwrap()))
```

Why does this code (and below) use `unwrap`? I think they should return an error rather than `panic`ing

---

I have another PR to address this. The interface of the Rewrite/transform Closures need to be changed.

https://github.com/apache/arrow-datafusion/pull/4318